Data/These.hs

```diff
@@ -194,6 +197,30 @@ catThat = mapMaybe justThat
 catThese :: [These a b] -> [(a, b)]
 catThese = mapMaybe justThese
 
+-- | Maybe map with Either like behavior on the left.
+mapMaybeThis :: Monoid c => (a -> Maybe b) -> These a c -> These b c
+mapMaybeThis f (This a) = maybe (That mempty) This (f a)
+mapMaybeThis f (These a c) = maybe (That c) This (f a)
+mapMaybeThis _ (That c) = That c
+
+-- | Maybe map with Either like behavior on the Right.
+mapMaybeThat :: Monoid c => (a -> Maybe b) -> These c a -> These c b
+mapMaybeThat _ (This c) = This c
+mapMaybeThat f (That a) = maybe (This mempty) That (f a)
+mapMaybeThat f (These c a) = maybe (This c) That (f a)
+
+-- | Maybe map with Either like behavior on both sides.
+mapMaybeThese :: (Monoid c, Monoid d) => (a -> Maybe c) -> (b -> Maybe d) -> These a b -> These c d
```

Comment:
So `mapMaybeThese` might turn `This` into `That` and `That` into `This` and `These` into `That` and `This`. Confusing. Give me some motivating examples (I'll add them to the haddocks then).

Reply:
This one I just added for completeness. I can remove it if you dont think its useful.